picard/browser/browser.py

```diff
@@ -25,115 +25,169 @@
+            self.server.server_close()
+            self.server = None
+            self.listen_port_changed.emit(self.port)
         else:
             log.debug("Browser integration inactive, no need to stop")
 
-    def _process_request(self):
-        conn = self.sender()
-        rawline = conn.readLine().data()
-        log.debug("Browser integration request: %r", rawline)
 
-        def parse_line(line):
-            orig_line = line
-            try:
-                line = line.split()
-                if line[0] == "GET" and "?" in line[1]:
-                    parsed = urlparse(line[1])
-                    args = parse_qs(parsed.query)
-                    if 'id' in args and args['id']:
-                        mbid = args['id'][0]
-                        if not mbid_validate(mbid):
-                            log.error("Browser integration failed: bad mbid %r", mbid)
-                            return False
-
-                        def load_it(loader):
-                            self.tagger.bring_tagger_front()
-                            loader(mbid)
-                            return True
-                        action = parsed.path
-                        if action == '/openalbum':
-                            return load_it(self.tagger.load_album)
-                        elif action == '/opennat':
-                            return load_it(self.tagger.load_nat)
-            except Exception as e:
-                log.error("Browser integration failed with %r on line %r", e, orig_line)
-                return False
-            log.error("Browser integration failed: cannot parse %r", orig_line)
-            return False
+class RequestHandler(BaseHTTPRequestHandler):
+
+    def do_OPTIONS(self):
+        origin = self.headers['origin']
+        if _is_valid_origin(origin):
+            self.send_response(204)
+            self.send_header('Access-Control-Allow-Origin', origin)
+            self.send_header('Access-Control-Allow-Methods', 'GET')
+            self.send_header('Access-Control-Allow-Credentials', 'false')
+            self.send_header('Access-Control-Allow-Private-Network', 'true')
+            self.send_header('Access-Control-Max-Age', 3600)
+            self.send_header('Vary', 'Origin')
+        else:
+            self.send_response(401)
+        self.end_headers()
 
+    def do_GET(self):
         try:
-            line = rawline.decode()
-            if parse_line(line):
-                conn.write(response(200))
+            self._handle_get()
+        except Exception:
+            log.error('Browser integration failed handling request', exc_info=True)
+            self._response(500, 'Unexpected request error')
+
+    def _handle_get(self):
+        parsed = urlparse(self.path)
+        args = parse_qs(parsed.query)
+        action = parsed.path
+
+        if action == '/':
+            self._response(200, SERVER_VERSION)
+        elif 'id' in args and args['id']:
```

Comment:
We should perhaps invert the logic here, checking valid actions first, and call `_load_mbid()` with the method to use, something like:

```python
        if action == '/':
            self._response(200, SERVER_VERSION)
        elif action == '/openalbum':
            self._load_mbid(tagger.load_album, args)
        elif action == ''/opennat':
             self._load_mbid(tagger.nat_album, args)
        else:
             self._response(400, 'Unknown action')
```

And do the parsing of `id` in `_load_mbid()`:

```python
def _load_mbid(self, loading_func, args):
    if 'id' in args and args['id']:
       mbid = args['id'][0]
       if not mbid_validate(mbid):
           self._response(400, '"id" is not a valid MBID.')
       else:
           tagger = QtCore.QCoreApplication.instance()
           def load_it(loader):
               tagger.bring_tagger_front()
               loader(mbid)
           to_main(load_it, loading_func)
           self._response(200, 'MBID "%s" loaded' % mbid)
    else:
        self._response(400, 'Missing parameter "id".')
```
 (indentation may be incorrect)